Fixes: #38554 - Host registration on Debian based systems fail because dpkg options are improperly formatted

Macleykun opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

The Why:
We noticed after upgrading foreman that we couldn't register Ubuntu hosts anymore. I created a forum post about it to see who has the issue aswell:
https://community.theforeman.org/t/host-registration-script-fails-with-dpkg-error-need-an-action-option/43619

The error:

```bash
here it goes wrong
+ apt-get -o Dpkg::Options::='--force-confdef' -o Dpkg::Options::='--force-confold' install -y subscription-manager
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following packages were automatically installed and are no longer required:
  linux-headers-5.15.0-140 linux-headers-5.15.0-140-generic linux-image-5.15.0-140-generic linux-modules-5.15.0-140-generic linux-modules-extra-5.15.0-140-generic
Use 'apt autoremove' to remove them.
The following additional packages will be installed:
  python3-subscription-manager
Suggested packages:
  python3-subscription-manager-doc
The following NEW packages will be installed:
  python3-subscription-manager subscription-manager
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/468 kB of archives.
After this operation, 2813 kB of additional disk space will be used.
dpkg: error: need an action option
```

The search:
Running that command in the terminal worked, also using it directly into the script worked.
A few users pointed out the quoting, but i didn't saw anything wrong with it.
One user also pointed to the PR that may have caused this issue: https://github.com/theforeman/foreman/pull/10436

The workaround:
I noticed if i installed the subscription-manager after the script broke, that it worked. So in our playbook i run the registration script, and if it fails i install the subscription-manager and then retry the registration script again.
But this felt wrong and i can imagine other people may have this issue aswell i though.

The solution:
After a busy week with other prioties i took the whole script and checked on quote's/formatting errors but didn't notice anything again.
Using set -x i did notice the same error again and with the users pointing out about missing a quote, i wonderd if the single quotes were causing the issue.
I did saw it used alot (single quote's) for the various distributions, but for debian these extra options were unique so i gave it a try.

```bash
here it goes wrong
+ apt-get -o Dpkg::Options::=--force-confdef -o Dpkg::Options::=--force-confold install -y subscription-manager
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following packages were automatically installed and are no longer required:
  linux-headers-5.15.0-140 linux-headers-5.15.0-140-generic linux-image-5.15.0-140-generic linux-modules-5.15.0-140-generic linux-modules-extra-5.15.0-140-generic
Use 'apt autoremove' to remove them.
The following additional packages will be installed:
  python3-subscription-manager
Suggested packages:
  python3-subscription-manager-doc
The following NEW packages will be installed:
  python3-subscription-manager subscription-manager
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/468 kB of archives.
After this operation, 2813 kB of additional disk space will be used.
Selecting previously unselected package python3-subscription-manager.
(Reading database ... 167795 files and directories currently installed.)
Preparing to unpack .../python3-subscription-manager_1.30.5-2_amd64.deb ...
Unpacking python3-subscription-manager (1.30.5-2) ...
Selecting previously unselected package subscription-manager.
Preparing to unpack .../subscription-manager_1.30.5-2_all.deb ...
Unpacking subscription-manager (1.30.5-2) ...
Setting up python3-subscription-manager (1.30.5-2) ...
Setting up subscription-manager (1.30.5-2) ...
Processing triggers for man-db (2.12.0-4build2) ...
Processing triggers for dbus (1.14.10-4ubuntu4.1) ...
Scanning processes...
Scanning linux images...

Running kernel seems to be up-to-date.

No services need to be restarted.

No containers need to be restarted.

No user sessions are running outdated binaries.

No VM guests are running outdated hypervisor (qemu) binaries on this host.
+ set +x
```

It worked! Because i only have an issue with this part of the script for Ubuntu hosts, i only changed to what i'm sure actually fixes the functionality and not break anything else (especially if i can't test it).

I hope this is something that can be merged relatively soon and published as a minor version upgrade for foreman. As at the moment the script won't work on the latest version of Foreman-Katello for Debian based hosts.

I'll ping @sbernhard who is the original author of the PR aswell as @stejskalleos  who merged it.
Happy friday and enjoy the weekend :) !

Macleykun committed:
Fixes #38554 - Host registration on Debian based systems fail because dpkg options are improperly formatted

m-bucher commented:
Thanks for the PR, @Macleykun.
Have you checked, if it works when you move the quotes like so:
```
apt-get -o 'Dpkg::Options::=--force-confdef' -o 'Dpkg::Options::=--force-confold' install -y subscription-manager
```
As far as the shell is concerned, it is not yet necessary to quote the command-line parameters (no special characters like whitespaces have been used, yet).
I assume that APT will understand it, if it gets `Dpkg::Options::=--force-confdef --something-else`, but the shell will need to know the whitespace is not a delimiter between cmdline-parameters for the `apt` command, but part of the value for the parameter `-o`.

Macleykun commented:
> Thanks for the PR, @Macleykun. Have you checked, if it works when you move the quotes like so:
> 
> ```
> apt-get -o 'Dpkg::Options::=--force-confdef' -o 'Dpkg::Options::=--force-confold' install -y subscription-manager
> ```
> 
> As far as the shell is concerned, it is not yet necessary to quote the command-line parameters (no special characters like whitespaces have been used, yet). I assume that APT will understand it, if it gets `Dpkg::Options::=--force-confdef --something-else`, but the shell will need to know the whitespace is not a delimiter between cmdline-parameters for the `apt` command, but part of the value for the parameter `-o`.

yup that works aswell:
![image](https://github.com/user-attachments/assets/d1caf359-7040-4e51-b9ed-b2a29ec3ef74)

if you want, i can adjust it to that if you want!
I can also change the PKG_MANAGER_UPGRADE aswel, even tho it wasn't part of that PR.

sbernhard commented:
@Macleykun you need to create a issue at projects.theforeman.org and reference it in your commit. See other commits of this project.

Macleykun commented:
@sbernhard rectified it in my last commit message + @m-bucher also changed the quotes + i also changed it for the upgrade arguments aswell.
I hope it's still acceptable to accept the PR as is
Issue i created for this: https://projects.theforeman.org/issues/38554

Macleykun commented:
@sbernhard heyo! Is there smth else in the PR you think needs adjusting?

m-bucher commented on app/views/unattended/provisioning_templates/snippet/pkg_manager.erb:
```diff
@@ -29,9 +29,9 @@ elif [ -f /etc/fedora-release -o -f /etc/redhat-release -o -f /etc/amazon-linux-
 elif [ -f /etc/debian_version ]; then
   PKG_MANAGER='apt-get'
-  PKG_MANAGER_INSTALL="${PKG_MANAGER} -o Dpkg::Options::='--force-confdef' -o Dpkg::Options::='--force-confold' install -y"
+  PKG_MANAGER_INSTALL="${PKG_MANAGER} -o 'Dpkg::Options::=--force-confdef' -o 'Dpkg::Options::=--force-confdef' install -y"
```
```suggestion
  PKG_MANAGER_INSTALL="${PKG_MANAGER} -o 'Dpkg::Options::=--force-confdef' -o 'Dpkg::Options::=--force-confold' install -y"
```

m-bucher reviewed (changes requested):
Hi, I think you have a copy-paste error there :-)

Macleykun commented:
> Hi, I think you have a copy-paste error there :-)

oof, good catch, really though i replaced the full line to avoid that from happening!

sbernhard commented:
can you please rebase & squash to 1 commit again?

Macleykun commented:
> can you please rebase & squash to 1 commit again?

Sure, I’ll have to see how I can do that in GitHub tho, but I believe it’s possible to merge and squash this pr aswell?

Macleykun commented:
> can you please rebase & squash to 1 commit again?

I managed to merge it into 1 commit! I'm very sure i have the right naming for the comimt and PR but i'm not sure why redmine still reports this as an issue.
https://github.com/theforeman/foreman/pull/10586/commits used this one as a reference, but all the parts i refer to are correct to the bug i'm trying to fix here: https://projects.theforeman.org/issues/38554

Macleykun commented:
ofcours i overlooked the : ... But the PR is now finnaly ready for merging! :D

Macleykun committed:
Merge branch 'theforeman:develop' into develop

Macleykun committed:
Revert "Fixes #38554 - Host registration on Debian based systems fail because dpkg options are improperly formatted"

This reverts commit 208a99ab4b73878468d823d834edb9513d26a9f7.

Macleykun committed:
revert previous commit
